### PR TITLE
Implement `FusedIterator` for `gen` block

### PR DESCRIPTION
**Pull request by ShoyuVanilla**
cc #117078

### CODE REVIEW

File: compiler/rustc_middle/src/traits/select.rs [hunk 1]
```diff
@@ -156,6 +156,10 @@ pub enum SelectionCandidate<'tcx> {
     /// generated for a `gen` construct.
     IteratorCandidate,
 
+    /// Implementation of an `FusedIterator` trait by one of the coroutine types
+    /// generated for a `gen` construct.
+    FusedIteratorCandidate,
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/traits/select.rs:161]: Just use `BuiltinCandidate { has_nested: false }`.

File: compiler/rustc_trait_selection/src/traits/util.rs [hunk 2]
```diff
@@ -323,6 +323,17 @@ pub fn iterator_trait_ref_and_outputs<'tcx>(
     (trait_ref, sig.yield_ty)
 }
 
+pub fn fused_iterator_trait_ref_and_outputs<'tcx>(
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/util.rs:326]: This is unnecessary -- fused iterators have no associated types and `FusedIterator` has no generic args.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs [hunk 3]
```diff
@@ -647,6 +647,41 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
         )
     }
 
+    fn consider_builtin_fused_iterator_candidate(
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:650]: `FusedIterator` has no associated items

File: compiler/rustc_middle/src/ty/instance.rs [hunk 4]
```diff
@@ -624,7 +624,9 @@ impl<'tcx> Instance<'tcx> {
                 hir::CoroutineKind::Desugared(hir::CoroutineDesugaring::Async, _)
             );
             hir::LangItem::FuturePoll
-        } else if Some(trait_id) == lang_items.iterator_trait() {
+        } else if Some(trait_id) == lang_items.iterator_trait()
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/instance.rs:627]: FusedIterator has no associated items

File: compiler/rustc_trait_selection/src/solve/assembly/mod.rs [hunk 5]
```diff
@@ -215,6 +215,14 @@ pub(super) trait GoalKind<'tcx>:
         goal: Goal<'tcx, Self>,
     ) -> QueryResult<'tcx>;
 
+    /// A coroutine (that comes from a `gen` desugaring) is known to implement
+    /// `FusedIterator<Item = O>`, where `O` is given by the generator's yield type
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/assembly/mod.rs:219]: No `<Item = O>`; `FusedIterator` has no associated items

File: compiler/rustc_trait_selection/src/solve/trait_goals.rs [hunk 6]
```diff
@@ -456,6 +456,13 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
         ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
     }
 
+    fn consider_builtin_fused_iterator_candidate(
+        ecx: &mut EvalCtxt<'_, 'tcx>,
+        goal: Goal<'tcx, Self>,
+    ) -> QueryResult<'tcx> {
+        Self::consider_builtin_iterator_candidate(ecx, goal)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/trait_goals.rs:463]: Don't do a nested call here. Just reimplement the logic entirely.

File: compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs [hunk 7]
```diff
@@ -313,6 +315,23 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
+    fn assemble_fused_iterator_candidates(
+        &mut self,
+        obligation: &PolyTraitObligation<'tcx>,
+        candidates: &mut SelectionCandidateSet<'tcx>,
+    ) {
+        let self_ty = obligation.self_ty().skip_binder();
+        if let ty::Coroutine(did, ..) = self_ty.kind() {
+            // gen constructs get lowered to a special kind of coroutine that
+            // should directly `impl FusedIterator`.
+            if self.tcx().coroutine_is_gen(*did) {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:327]: Use let-chains to make this into one `if`

**Review by compiler-errors — CHANGES_REQUESTED** [review]

**Comment by ShoyuVanilla** [comment]
I've done errornous implementation out of some misconceptions 😅
Will fix these soon. Sorry

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
🆒 

r? compiler-errors @bors r+ rollup

**Pull request merged**